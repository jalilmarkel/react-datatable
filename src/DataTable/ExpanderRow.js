import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components';

// Make "data" available on our any child component
// eslint-disable-next-line arrow-body-style
const renderChildren = (children, data) => {
  return Children.map(children, child => cloneElement(child, { data }));
};

const ExpanderRowStyle = styled.div`
  width: 100%;
  box-sizing: border-box;
  background-color: ${props => props.theme.expander.backgroundColor};
`;

const ExpanderRow = ({
  data,
  children,
}) => (
  <ExpanderRowStyle>
    {renderChildren(children, data)}
  </ExpanderRowStyle>
);

ExpanderRow.propTypes = {
  data: PropTypes.object,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

ExpanderRow.defaultProps = {
  data: {},
  children: null,
};

export default withTheme(ExpanderRow);
