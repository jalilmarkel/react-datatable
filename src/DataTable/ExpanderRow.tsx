import * as React from 'react';
import styled, { CSSObject } from 'styled-components';
import { ComponentProps, ExpandableRowsComponent } from './types';

const ExpanderRowStyle = styled.div<{
	extendedRowStyle: CSSObject;
}>`
	width: 100%;
	box-sizing: border-box;
	${({ theme }) => theme.expanderRow.style};
	${({ extendedRowStyle }) => extendedRowStyle};
`;

type ExpanderRowProps<T> = {
	data: T;
	component: ExpandableRowsComponent;
	extendedRowStyle: CSSObject;
	extendedClassNames: string;
	componentProps: ComponentProps;
};

function ExpanderRow<T>({
	data,
	component,
	componentProps,
	extendedRowStyle,
	extendedClassNames,
}: ExpanderRowProps<T>): JSX.Element {
	const ExpandableComponent = component;
	// we need to strip of rdt_TableRow from extendedClassNames
	const classNamesSplit = extendedClassNames.split(' ').filter(c => c !== 'rdt_TableRow');
	const classNames = ['rdt_ExpanderRow', ...classNamesSplit].join(' ');

	return (
		<ExpanderRowStyle className={classNames} extendedRowStyle={extendedRowStyle}>
			<ExpandableComponent data={data} {...componentProps} />
		</ExpanderRowStyle>
	);
}

export default React.memo(ExpanderRow);
