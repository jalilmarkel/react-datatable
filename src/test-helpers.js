import React from 'react';
import { render } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';
import theme from './themes/default';

export const renderWithTheme = (tree, ...args) =>
  render(<ThemeProvider theme={theme}>{tree}</ThemeProvider>, ...args);

export function syntheticEvent(a) {
  return ({
    target: a,
    preventDefault: jest.fn(),
    stopPropagation: jest.fn(),
  });
}
