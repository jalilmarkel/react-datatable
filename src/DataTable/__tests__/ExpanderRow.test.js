import 'jest-styled-components';
import React from 'react';
import { cleanup } from 'react-testing-library';
import ExpanderRow from '../ExpanderRow';
import { renderWithTheme } from '../../test-helpers';

afterEach(cleanup);

test('component <ExpanderRow /> should render correctly', () => {
  const { container } = renderWithTheme(<ExpanderRow />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <ExpanderRow numColumns /> should render correctly', () => {
  const { container } = renderWithTheme(<ExpanderRow numColumns={5} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <ExpanderRow /> with children should render correctly and have a data prop', () => {
  const { container } = renderWithTheme(<ExpanderRow data={{ name: 'morty' }}><div>Some Expander Content</div></ExpanderRow>);

  expect(container.firstChild).toMatchSnapshot();
});
