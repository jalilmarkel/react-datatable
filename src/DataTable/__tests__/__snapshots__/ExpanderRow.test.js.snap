// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpanderRow /> should render correctly 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  background-color: transparent;
}

<div
  class="c0"
/>
`;

exports[`<ExpanderRow /> with children should render correctly and have a data prop 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  background-color: transparent;
}

<div
  class="c0"
>
  <div
    data="[object Object]"
  >
    Some Expander Content
  </div>
</div>
`;

exports[`<ExpanderRow numColumns /> should render correctly 1`] = `
.c0 {
  width: 100%;
  box-sizing: border-box;
  background-color: transparent;
}

<div
  class="c0"
/>
`;
